### PR TITLE
Added forwarded messages

### PR DESCRIPTION
**Pull request by DubyaDude**
# Summary
Added forwarded messages deserialization support as per discord message docs: https://discord.com/developers/docs/resources/message

### CODE REVIEW

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 1]
```diff
@@ -389,6 +397,11 @@ private IMention[] GetMentions()
             mentions.Add(new RepliedUserMention()); // Return null to allow all mentions
         }
 
+        if (this.MessageSnapshots is not null && this.mentionedUsers.Any(r => this.MessageSnapshots.Any(m => r.Id == m.Message?.Author?.Id)))
+        {
+            mentions.Add(new RepliedUserMention()); // Return null to allow all mentions
+        }
+
```
Comment (DubyaDude) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:404]: I'm unsure if this is needed, couldn't quite understand it. Modeled it after ReferencedMessage. Let me know if I should remove this.

File: DSharpPlus/Entities/Channel/Message/DiscordMessageSnapshot.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+/// <summary>
+/// Represents a Discord message snapshot.
+/// </summary>
+public class DiscordMessageSnapshot
+{
+    /// <summary>
+    /// Gets the message object for the message snapshot.
+    /// </summary>
+    [JsonProperty("message", NullValueHandling = NullValueHandling.Ignore)]
+    public DiscordMessage Message { get; set; }
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/Message/DiscordMessageSnapshot.cs:14]: this needs to be its own type, the DiscordMessage object assumes it always has an ID, which this doesn't

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs [hunk 3]
```diff
@@ -300,6 +300,12 @@ public IReadOnlyList<DiscordMessageSticker>? Stickers
     [JsonProperty("referenced_message", NullValueHandling = NullValueHandling.Ignore)]
     public DiscordMessage? ReferencedMessage { get; internal set; }
 
+    /// <summary>
+    /// Gets the message object for the referenced message
+    /// </summary>
+    [JsonProperty("message_snapshots", NullValueHandling = NullValueHandling.Ignore)]
+    public DiscordMessageSnapshot[]? MessageSnapshots { get; internal set; }
```
Comment (VelvetToroyashi) [inline DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:307]: This should be IReadOnlyList<DiscordMessageSnapshot> to align with other public APIs

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs (hunk 1 above)
Comment (VelvetToroyashi) [reply DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:404]: Based on what I can understand of the code, and how forwards currently function, forwards cannot ping, so this is probably incorrect

**Review by VelvetToroyashi — CHANGES_REQUESTED** [review]
There's more, but I can't really speak for certain about it, as the code more confuses me than anything

File: DSharpPlus/Entities/Channel/Message/DiscordMessage.cs (hunk 1 above)
Comment (DubyaDude) [reply DSharpPlus/Entities/Channel/Message/DiscordMessage.cs:404]: Sounds like this block can just be removed then?

File: DSharpPlus/Entities/Channel/Message/DiscordMessageReferenceType.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+namespace DSharpPlus.Entities;
+
+
+/// <summary>
+/// Represents the type of a message reference.
+/// </summary>
+public enum DiscordMessagReferenceType : int
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/Message/DiscordMessageReferenceType.cs:7]: typo

File: DSharpPlus/Entities/Channel/Message/DiscordMessageSnapshot.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+using Newtonsoft.Json;
+
+namespace DSharpPlus.Entities;
+
+/// <summary>
+/// Represents a Discord message snapshot.
+/// </summary>
+public class DiscordMessageSnapshot
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/Message/DiscordMessageSnapshot.cs:8]: sealed?

**Review by akiraveliara — CHANGES_REQUESTED** [review]
the typo in the type name should be fixed, other than that lgtm

**Review by akiraveliara — APPROVED** [review]

**Comment by Plerx2493** [comment]
Tested?

**Comment by DubyaDude** [comment]
> Tested?

Yes

**Pull request merged**